### PR TITLE
Dependency resolution - Do not pin candidates on first selection

### PR DESCRIPTION
**Pull request by vphilippon**
Trying to address #875, #993, #1060

What this PR still needs:
- [x] A test to reproduce the issue where pre-pinning candidate causes the resolution to fail.
- [x] A test that covers what the initial patch from Kenneth intended to fix.

Details can be found on this comment: https://github.com/kennethreitz/pipenv/issues/875#issuecomment-337717817

### CODE REVIEW

**Comment by vphilippon** [comment]
Note to self: need to test with `records[pandas]`, it must result in a lock with pandas included, which was the original intended patch.

File: tests/test_pipenv.py [hunk 1]
```diff
@@ -820,6 +820,27 @@ def test_complex_deps_lock_and_install_properly(self):
             c = p.pipenv('install')
             assert c.return_code == 0
 
+    @pytest.mark.lock
+    @pytest.mark.requirements
+    @pytest.mark.complex
+    def test_complex_lock_changing_candidate(self):
+        # The requests candidate will change from latest to <2.12.
```
Comment (erinxocon) [inline tests/test_pipenv.py]: Do we want to check the version numbers in the lock file to see if it reports <2.12

**Comment by erinxocon** [comment]
@vphilippon Thanks so much for this!  This will be a great fix for deep dependency resolution.  I think the tests are great with one comment.  I'll let @kennethreitz or @nateprewitt take a look as well.  I'm not as familiar with the pip internals, but this does appear to work for me when I checkout this PR and test.

File: tests/test_pipenv.py (hunk 1 above)
Comment (vphilippon) [reply tests/test_pipenv.py]: That would be good idea to make sure that we're resolving *and* giving a valid result, indeed!

Comment (vphilippon) [reply tests/test_pipenv.py]: Updated the test

**Comment by erinxocon** [comment]
@vphilippon Weird the app veyor test for python34 took like 20 minutes...App Veyor can be odd

**Comment by vphilippon** [comment]
@erinxocon Yeah, the flaky CI does worry me a bit, as I rely on it for many tests (I have a bunch of issues locally with pipenv, which means I can't run all the tests locally before pushing)

So the Appveyor tests passed, that's nice. That one Travis job that is failing (Py3.4) bugs me, as I relaunched it, to no avail. I'll give it a 3rd try.

**Comment by erinxocon** [comment]
@vphilippon Huh the 3.4 test on travis still fails.  It looks like all of the tests run ok, but it just stalls.  Do we have any tests that would leave a hanging resource or pointer open?

**Comment by erinxocon** [comment]
@vphilippon It looks like all of the tests pass for python34, but it's hanging, despite all of the tests passing I don't think we can merge until we resolve this.  Any ideas? I'm not seeing something obvious.

**Comment by vphilippon** [comment]
@erinxocon No idea from my side either. I'm not sure what could be hanging the tests. That does need investigation, although I'm not sure when I'll be able to look into this.

**Comment by nateprewitt** [comment]
I bounced it to get a fresh set of tests run. Let's see if this is tied to host state or not.

**Comment by vphilippon** [comment]
IT WORKED!

**Comment by nateprewitt** [comment]
Looking good, thanks for getting this put together @vphilippon!

**Pull request merged**